harvester/metadata.py

```diff
@@ -0,0 +1,278 @@
+                    logger.debug(
+                        "%d/%d records parsed, %.2fs", *(i, row_count, time.time() - t0)
+                    )
+                    t0 = time.time()
+
+                # init website metadata dictionary with data already known from WACZ files
+                metadata = {
+                    "url": row.url,
+                    "cdx_warc_filename": row.filename,
+                    "cdx_title": row.title,
+                    "cdx_offset": row.offset,
+                    "cdx_length": row.length,
+                }
+
+                # augment with metadata parsed from the website's HTML content
+                html_content = wacz_client.get_website_content(
+                    row.filename, row.offset, decode=True
+                )
+                html_metadata = self.get_html_content_metadata(html_content)
+                metadata.update(html_metadata)
+
+                # augment again with data parsed from, and including, HTML fulltext
+                metadata.update(
+                    self.parse_fulltext_fields(
+                        row.text, include_fulltext=include_fulltext
+                    )
+                )
+
+                all_metadata.append(metadata)
+
+            # create dataframe from all dictionaries
+            websites_metadata_df = pd.DataFrame(all_metadata)
+
+            # replace NaN with python None
+            websites_metadata_df = websites_metadata_df.where(
+                pd.notna(websites_metadata_df), None
+            )
+
+            # init instance of CrawlMetadataRecords and cache
+            self._websites_metadata = CrawlMetadataRecords(websites_metadata_df)
+
+            return self._websites_metadata
+
+    @classmethod
+    def get_html_content_metadata(cls, html_content: str | bytes) -> dict:
+        """Extract structured metadata from the website HTML.
+
+        This method extracts values for HTML tags defined in HTML_CONTENT_METADATA_TAGS,
+        which are mostly aligned with expected Open Graph Protocol (OGP) HTML elements.
+        """
+        soup = BeautifulSoup(html_content, "html.parser")
+        tags = {}
+        for og_tag_name in cls.HTML_CONTENT_METADATA_TAGS:
+            og_tag = soup.find("meta", property=og_tag_name)
+            if og_tag is None:
+                continue
+            if content := og_tag.get("content"):  # type: ignore[union-attr]
+                content_stripped = content.strip()  # type: ignore[union-attr]
+                if content_stripped != "":
+                    og_tag_name_friendly = og_tag_name.replace(":", "_")
+                    tags[og_tag_name_friendly] = content_stripped
+        return tags
+
+    def _remove_fulltext_whitespace(self, fulltext: str) -> str:
+        """Cleanup fulltext as provided by Browsertrix crawl."""
+        fulltext = fulltext.replace("\n", " ")
+        fulltext = fulltext.replace("\r", " ")
+        fulltext = fulltext.replace("\t", " ")
```

Comment:
I don't think `<str>.replace` modifies in place, but this could be changed.

For context, I think this fulltext handling will be the first thing to get improved after some rounds of testing and feedback.  So, it's pretty naive and simple now.  But will update to a chained method approach.